### PR TITLE
Fix different precedence binary expression when inlining

### LINKED ISSUE #2823: Extra indent in a 'condition && <JSX>' expression
**Prettier 1.7.0**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22const%20avatar%20%3D%20(hasAvatar%20%7C%7C%20showPlaceholder)%20%26%26%20(%20%3CGravatar%20user%3D%7Bauthor%7D%20size%3D%7Bsize%7D%20%2F%3E%20)%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A50%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
# Options (if any):
--print-width=50
```

**Input:**
```jsx
const avatar = (hasAvatar || showPlaceholder) && ( <Gravatar user={author} size={size} /> )
```

**Output:**
```jsx
const avatar = (hasAvatar || showPlaceholder) && (
    <Gravatar user={author} size={size} />
  );
```

**Expected behavior:**
```jsx
const avatar = (hasAvatar || showPlaceholder) && (
  <Gravatar user={author} size={size} />
);
```

This is a regression since 1.6.1 -- that version formatted this code correctly.

### PR DESCRIPTION
**Pull request by duailibe**
Fixes #2823

With the fix in #2770 I ended up introducing a bug when the chain has
different precedences.

### CODE REVIEW

File: tests/binary-expressions/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -208,13 +253,48 @@ prevState = prevState ||
     selectedCatalog: null
   };
 
+prevState =
+  prevState ||
```
Comment (suchipi) [inline tests/binary-expressions/__snapshots__/jsfmt.spec.js.snap]: Why does this one break the `prevState` onto the next line, while the one on line 266 doesn't?

Comment (duailibe) [reply tests/binary-expressions/__snapshots__/jsfmt.spec.js.snap]: @suchipi As I understand it, the object is a subexpression of the right side group. So `shouldInlineLogicalExpression` is false for the outermost group, but true for the inner group in the right side.

See [playground](https://prettier.io/playground/#%7B%22content%22%3A%22const%20a%20%3D%20longlonglonglonglonglonglong%20%7C%7C%20longlonglonglonglonglonglong%3B%5Cn%5Cnconst%20a%20%3D%20longlonglonglonglonglonglong%20%7C%7C%20longlonglonglonglonglonglong%20%7C%7C%20%7B%5Cn%20%20prop%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A50%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)

**Comment by lydell** [comment]
This looks good to me, except that I don't really understand which test corresponds to the case reported in #2823?

**Comment by duailibe** [comment]
@lydell It just follows the same rules for inlining arrays and objects.

I'll add specific test cases for JSX

**Comment by duailibe** [comment]
I added a new test case and rebased to master

File: tests/binary-expressions/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -145,6 +145,37 @@ if (
...
+  </div>
+);
+
+const user =
+  renderedUser ||
+  (shouldRenderUser && (
+    <div>
+      <User name={this.state.user.name} age={this.state.user.age} />
+    </div>
+  ));
```
Comment (suchipi) [inline tests/binary-expressions/__snapshots__/jsfmt.spec.js.snap]: At first glance, this double-paren looked kinda weird, but after thinking it over, I think I actually prefer it to any of the alternatives

**Pull request merged**